File: Hash/133.Clone_Graph/133.Clone_Graph.cpp
//Solution 1
class Solution {
public:
    Node* cloneGraph(Node* node) {
        unordered_map<Node*, Node*> mapping;
        queue<Node*> q;
        q.push(node);
        
        while(!q.empty()) {
            Node *temp = q.front(); q.pop();
            mapping[temp] = new Node(temp->val, {});
            for(auto neighbor: temp->neighbors){
                if(!mapping.count(neighbor)) q.push(neighbor);
            }
        }
        
        for (auto ele: mapping) {
            for(auto neighbor: ele.first->neighbors) {
                ele.second->neighbors.push_back(mapping[neighbor]);
            }
        }
        return mapping[node];
    }
};

//Solution 2
class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (!node) return NULL;
        if (mapping.count(node)) return mapping[node];
        
        mapping[node] = new Node(node->val, {});
        for(auto neighbor: node->neighbors) {
            mapping[node]->neighbors.push_back(cloneGraph(neighbor));
        }
        return mapping[node];
    }
    
    unordered_map<Node*, Node*> mapping;
};

File: Hash/133.Clone_Graph/133.Clone_Graph.py
# BFS
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node :
            return
        copynode = Node(node.val, [])
        dic = {node: copynode}
        queue = collections.deque([node])
        
        while queue :
            temp = queue.popleft();
            for neighbor in temp.neighbors :
                if neighbor not in dic :
                    copynode = Node(neighbor.val, [])
                    dic[neighbor] = copynode
                    dic[temp].neighbors.append(copynode)
                    queue.append(neighbor)
                else :
                    dic[temp].neighbors.append(dic[neighbor])
        
        return dic[node]

# DFS
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        dic = {}
        def DFS(node) :
            if not node :
                return 
            if node in dic :
                return dic[node]
            dic[node] = Node(node.val, [])
            
            for neighbor in node.neighbors :
                dic[node].neighbors.append(DFS(neighbor))
            
            return dic[node]
        
        return DFS(node)